README.md
# Open Bodem Index Calculator (OBIC)

<!-- badges: start -->
[![Lifecycle: stable](https://img.shields.io/badge/lifecycle-stable-green.svg)](https://www.tidyverse.org/lifecycle/#stable)
[![R-CMD-check](https://github.com/springgbv/Open-Bodem-Index-Calculator/workflows/R-CMD-check/badge.svg)](https://github.com/springgbv/Open-Bodem-Index-Calculator/actions)
[![Codecov test coverage](https://codecov.io/gh/springgbv/Open-Bodem-Index-Calculator/branch/master/graph/badge.svg)](https://codecov.io/gh/springgbv/Open-Bodem-Index-Calculator?branch=master)
<!-- badges: end -->

This R package can be used to calculate the Open Bodem Index (OBI). 
The OBI is a tool that evaluate the soil of agricultural fields based on four main criteria: chemical, physical, biological and management. 
These four criteria consist of more than 21 indicators to get a comprehensive picture of the soil. 
More information about the Open Bodem Index can be found at [Open Bodem Index](https://www.openbodemindex.nl).

## Installation

OBIC can be installed from GitHub using `remotes`. This will install the current development version from master branch. Stable versions can found at `releases`.
```
remotes::install_github("AgroCares/Open-Bodem-Index-Calculator")
library("OBIC")
```
## Documentation
Documentation of the R functions in the OBIC package can be found [here](https://agrocares.github.io/Open-Bodem-Index-Calculator/)

## More information
The [Open Bodem Index](https://www.openbodemindex.nl) (OBI) is a collaboration between Wageningen UR, NMI and FarmHack in collaboration with agricultural experts. 
The Open Bodem Index Calculator is a tool used by the OBI and developed by [NMI](https://nmi-agro.nl/).

![Logo of NMI](https://media.licdn.com/dms/image/C560BAQEYGcm4HjNnxA/company-logo_200_200/0?e=2159024400&v=beta&t=u40rJ7bixPWB2SAqaj3KCKzJRoKcqf0wUXCdmsTDQvw)


NEWS.md
# Changelog OBIC

## UNRELEASED
### Added
* Adds GitHub Action to run R-CMD-Check for changes on `master` and `development`
* Adds coverage of unit tests

### Fixed
* Fixes ASCII signs at `crops.obic` and `column_description.obic`

## Version 1.1.1 2021-05-21
### Added
* table nema.crop.rot.obic added
* ppr_bouwplan_tables.r in dev/scripts
* merge_nema_tables.r in dev/scripts

## Version 1.1.0 2021-05-03
### Added
* argument B_GWL_GHG, B_GWL_GLG and B_Z_TWO to obic_field and obic_field_dt
* function calc_workability is active now

### Changed
* updated documentation, weight.obic for workability
* update calc_workability
* update column_descriptons_obic.Rdata

## Version 1.0.0 2021-04-28
### Added
* column crops_season to crops.obic
* column scientific_names to crops.obic
* preparation crops.obic in dev
* function format_gwt to reformat input for groundwater table 
* function format_soilcompaction to reformat input for subsoil compaction
* season.obic added as data.table
* new function and tests for workability indicator
* nema.obic added as data.table
* new function and tests for nematode indicator
* preparation nema.obic in dev
* preparation season.obic in dev
* obic_field is added to run obic score for one field
* ppr_column_description plus column_descriptons_obic.Rdata
* ppr_maatregel plus recom_obic.Rdata
* ppr_weights plus weights.obic
* I_P_DS and I_P_WS added (drought and wetness stress separately)
* add_management added to estimate default values for measures when not given 
* set of eight I_M_XX indicators for measures are added
* set of nine I_X_BCS indicators are added (optional)
* new M_codes added: M_LIME, M_NONINVTILL, M_SSPM, M_SOLIDMANURE,M_STRAWRESIDUE, M_MECHWEEDS, M_PESTICIDES_DST, issue #107
* obic_field_dt to run obic_field() for a data.table input 

### Changed
* calculation PBI updated for grassland
* PBI input is flexible for both grassland and arable
* calculation SLV updated
* update element_names for A_P_CC, A_P_AL in phosphate_availability.R
* update element_names for A_S_RT, A_SOM_LOI, B_AER_CBS, B_SOILTYPE_AGR in sulpher.R
* update element_names for B_GWL_CLASS in wateravailability.R
* update element_names for B_GWL_CLASS and B_AER_CBS in nretention.R
* correct weighing for S_C_A
* all element names are updated to english ones
* weighing for score calculated is changed into a more generic approach
* structure for running obi is changed into one function
* weights.obic is extended and used to distinguish for relevant/non-relevant indicators
* grass_age function is updated: count actual years
* BCS is optional input, and when given it overwrite risk indicators compaction / aggregate stability
* M codes for no advice, issue #62
* directory dev is updated and cleaned, issue #104
* lower and upper limits for soil properties updated
* allow unknown B_HELP_WENR in checkmate water_availability, issue 65
* very low Mg indicator values for maize are changed, plus tests updated
* very low Mg indicator values for grassland on clay and peat, issue 84
* evaluation and calculation for Cu are updated

### Fixed
* setorder in winderodibility


## Version 0.11.1 2020-09-02
### Changed
* The water holding capacity is now between 0 and 1 (as theta should be) and uses new evaluation function #67
* Lower range of `D_NLV` is set to -30
* Upper range of `D_MG` is set to 1000
* Range of `D_P_DU` is set between 0 and 1

### Fixed
* Fixed the use of `B_LG_CBS` at sulphur
* `ID` can now also be character and does not have to be numeric #66


## Version 0.11.0 2020-01-16
### Added
* Adds two other pedotransfer functions (PTFs) to calculate water retention parameters #53

### Changed
* Optimal pH values of grassland are now the middle value of the class `good` #55
* pF of field capacity is changed from 2.2 to 2.0 #53
* Bulk density in the default PTF is now aggregated into 1 continous function #53
* Parameters of evaluation for PAW are adjusted #53
* Parameters for the evaluation of sealing are adjusted #57
* Changes the formula to calculate nitrogen leaching and run-off #58
* Parameters for the evaluation of nretention are adjusted #58
* Paramaters and funs for CEC are updated and split for soil fertility and aggregate stability #59
* Parameters for the evaluation of Cu availability are adjusted #59
* potassium index is multiplied so that the evaluation is equal for grass and maize #60

### Fixed
* Improved the function for crumbleability to match the factsheet #54 #56
* Fix error in calc_potassium_availability #59

## Version 0.10.0 2019-12-13
### Added
* Adds the evaluation of measures #37
* Adds giving recommendations based on the OBI score #37
* A relative score per reference group is added #31
* Adds the leaching of nitrogen to surface water and groundwater #17
* Adds a table with the description of the columns and a vignette about it #46
* Adds scoring for environmental performance #31

### Changed
* Update of crumbeability groups in `crops.obic`
* Maximum value of `A_MN_CC` is increased from 30000 to 250000
* The code for evaluating  in `ind_crumbeability` is simplified
* M_M10 user input changed in `D_CP_RUST` input, plus comments added #29
* new table with weighing factors included 'weight.obic' #32
* weighing factors updated in obic_indicators.R #32
* Improved the zinc indicator function #39

tests/testthat/test-workability.R
test_that("calc_workability works", {
  expect_equal(
    calc_workability(
      A_CLAY_MI = c(15.6,13.6, 4.3, 22.6, 1.9, 2.9, 3.1, 4.3, 15.6, 1.9),
      A_SILT_MI = c(16.7,30.5, 11.8, 36.6, 9.2, 8.6, 10.6, 11.8, 16.7, 9.2),
      B_LU_BRP = c(233, 234, 236, 256, 259, 265, 265, 317, 2014, 259),
      B_SOILTYPE_AGR = c('zeeklei','zeeklei', 'dekzand','zeeklei', 'dekzand', 'dekzand', 'veen', 'dekzand', 'zeeklei', 'maasklei'),
      B_GWL_GLG = c(173,139, 106, 144, 115, 113, 42, 106, 173, 115),
      B_GWL_GHG = c(21, 18, 62, 70, 49, 81, 9, 62, 21, 49),
      B_GWL_ZCRIT = c(400, 400, 400, 400, 400, 400, 400, 400, 400, 400)
    ),
    expected = c(0.57,0.45,0.54,0.84,0.80, 1, 0.00,0.76,0.66, 0.66),
    tolerance = .01
  )
})

test_that("ind_workability works", {
  expect_equal(
    ind_workability(
      D_WO = c(0, 0.75, 1, 0, 0.5, 1),
      B_LU_BRP = c(256, 256, 256, 265, 265, 265)
    ),
    expected = c(0, 0.5, 1, 0, 0.5, 1),
    tolerance = .015
  )
})

test_that('diffrenttiation maize on sand vs mais on clay workabilty season works',{
  expect_true(calc_workability(
    A_CLAY_MI = c( 1.9),
    A_SILT_MI = c(9.2),
    B_LU_BRP = c(259),
    B_SOILTYPE_AGR = c( 'dekzand'),
    B_GWL_GLG = c(115),
    B_GWL_GHG = c(49),
    B_GWL_ZCRIT = c(400)
  )>
    calc_workability(
      A_CLAY_MI = c( 1.9),
      A_SILT_MI = c(9.2),
      B_LU_BRP = c(259),
      B_SOILTYPE_AGR = c( 'zeeklei'),
      B_GWL_GLG = c(115),
      B_GWL_GHG = c(49),
      B_GWL_ZCRIT = c(400)
    ))
})


R/grass_age.R
#' Calculate the average age of the grass
#' 
#' This function calculates the average age of the grass
#' 
#' @param ID (numeric) The ID of the field
#' @param B_LU_BRP (numeric) The crop code (gewascode) from the BRP
#' 
#' @import data.table
#' 
#' @export
calc_grass_age <- function(ID, B_LU_BRP) {
  
  all_count = crop_category = crop_code = grass = grass_age = grass_count = this_id = NULL
  
  # Load data
  crops.obic <- as.data.table(OBIC::crops.obic)
  setkey(crops.obic, crop_code)
  
  # Check inputs
  arg.length = max(length(ID), length(B_LU_BRP))
  checkmate::assert_atomic_vector(ID, len = arg.length)
  checkmate::assert_numeric(B_LU_BRP, len = arg.length)
  checkmate::assert_subset(B_LU_BRP, choices = unique(crops.obic$crop_code), empty.ok = FALSE)
  
  # Collect the data into a table
  dt <- data.table(
    this_id = 1:arg.length,
    ID = ID,
    B_LU_BRP = B_LU_BRP
  )
  
  # merge with crops.obic to get crop_category
  dt <- merge(dt, crops.obic[, list(crop_code, crop_category)], by.x = "B_LU_BRP", by.y = "crop_code")
  setorder(dt, this_id)
  
  # Calculate the age of the grassland
  dt[, grass := ifelse(crop_category == "grasland", 1, 0)]
  dt[, grass_age := rev(cumsum(grass)),by = rleid(ID, grass)]
 
  # extract grass age 
  value <- dt[, grass_age]
  
  # return grass age
  return(value)
}


DEV/scripts/ppr_weights.R
# make weights.obic table
# this table evaluates for each indicator whether its applicable yes or no
# applicability is currently differentiated for land use categories and soil type (peat and non-peat)
# in future this one can be used to differentiate between countries too
# weights are either 1 (applicable) or -1 (not applicable)

  # load library
  library(data.table)

  # load csv file
  weight.obic <- fread("dev/data/weight_obic.csv")

  # load the file with all indicators
  obic.ind <- as.data.table(readRDS('dev/data/column_description_obic.rds'))
  
  # which indicators should be present
  colsd <- obic.ind[type=="indicator",column]
  
  # which are present in weights
  cols <- unique(weight.obic$indicator)
  
  # which are missing
  cols.mis <- colsd[!colsd %in% cols]
  
  # which are superfluous
  cols.super <- cols[!cols %in% colsd]
  
  # checks
  if(length(cols.mis)==0){
    print('all indicators are also in column description')
  } else {
      print(paste0('indicators are missing: ',paste0(cols.mis,collapse=', ')))
  }
  if(length(cols.super)==0) {
    print('all indicators in weights.obic are also in column description')
  } else {
      print(paste0('some indicators are in weights.obic but do not have a column description: ', paste0(cols.super, collapse= ', ')))
  }
  
save(weight.obic, file = "data/weight_obic.RData")


DEV/scripts/ppr_season_obic.R
# Script to make season.obic table
 
  # require package
  library(data.table)

  # Table is based on Tabel 2 in Huinink (2018) and links to crops.obic$crop_season
  # Not all landuses in the original table appear in brp and some of the brp crops have been assigned to "overig" and could be reclassified later.
  season.obic <- fread('dev/data/season_obic.csv')
  
  # remove column numbers
  season.obic <- season.obic[,.(landuse, req_days_pre_glg, req_days_post_glg, total_days, derving)]
  
  # add soiltype columns to be molten based on soils.obic$soiltype.m
  season.obic <- season.obic[, zand := 'zand']
  season.obic <- season.obic[, klei := 'klei']
  season.obic <- season.obic[, veen := 'veen']
  season.obic <- season.obic[, loess := 'loess']
  
  # melt soiltype
  season.obic <- melt(season.obic, measure.vars = c('zand', 'klei', 'veen', 'loess'), variable.name = 'soiltype.m')
  
  # remove value column (only column names were really needed in melting)
  season.obic <- season.obic[,.(landuse, req_days_pre_glg, req_days_post_glg, total_days, derving, soiltype.m)]
  
  # set req days post glg for mais on sand/loess to match 1 october and other soils to 20 october
  season.obic <- season.obic[landuse == 'snijmais',req_days_post_glg := 
                               fifelse(soiltype.m %in% c('zand', 'loess'), 274-227, 293-227)] # where numbers are doy resulting in number of days after glg (doy = 227)
 
  # set negative required days to 0 but preserve this data in csv
  season.obic <- season.obic[req_days_post_glg < 0,req_days_post_glg := 0]
  
  # ensure that total days equal to or larger than required days pre and/or post glg
  season.obic <- season.obic[total_days<(req_days_pre_glg+req_days_post_glg)|
                               total_days<req_days_pre_glg|
                               total_days<req_days_post_glg,
                             total_days := req_days_post_glg+req_days_pre_glg]
  
  # save table
  usethis::use_data(name = season.obic, overwrite = TRUE)
  # save as csv for tracking
  write.csv(season.obic_orig, 'DEV/data/season.obic.csv')


R/workability.R
#' Calculate indicator for workability 
#'
#' This function calculates the workability of soils, given as a value of relative season length between 0 and 1. 
#' A relative season length of 1 indicates that the water table is sufficiently low for the soil to be workable for the entire growing season required by the crop.
#' The required ground water table for workability is determined by soil type and soil properties. Hydrological variables determine the groundwater table for each day of the year.
#' The option calcyieldloss allows for calculation of yield loss based on the relative season length, differentiating in yield loss between six groups of crops
#' Based on Huinink (2018)
#' 
#' @param A_CLAY_MI (numeric) The clay content of the soil (\%)
#' @param A_SILT_MI (numeric) The silt content of the soil (\%)
#' @param B_LU_BRP (numeric) The crop code from the BRP
#' @param B_SOILTYPE_AGR (character) The agricultural type of soil
#' @param B_GWL_GLG (numeric) The lowest groundwater level averaged over the most dry periods in 8 years in cm below ground level
#' @param B_GWL_GHG (numeric) The highest groundwater level averaged over the most wet periods in 8 years in cm below ground level
#' @param B_GWL_ZCRIT  (numeric) The distance between ground level and groundwater level at which the groundwater can supply the soil surface with 2mm water per day (in cm)
#' @param calcyieldloss (boolean) whether the function includes yield loss, options: TRUE or FALSE (default).
#' 
#' @import data.table
#' 
#' @references Huinink (2018) Bodem/perceel geschiktheidsbeoordeling voor Landbouw, Bosbouw en Recreatie. BodemConsult-Arnhem
#'  
#' @export
calc_workability <- function(A_CLAY_MI, A_SILT_MI, B_LU_BRP, B_SOILTYPE_AGR, 
                             B_GWL_GLG, B_GWL_GHG, B_GWL_ZCRIT,
                             calcyieldloss = FALSE) {
  
  # define variables used within the function
  id =crop_code = soiltype = landuse = crop_waterstress = crop_season = NULL
  soiltype.m = spring_depth =  gws_sub_workingdepth = NULL
  early_season_day_deficit = late_season_day_deficit = NULL
  req_days_pre_glg = req_days_post_glg = total_days = NULL
  derving = yield = yl = NULL
  
  # parameters related to required depth (rd): required depth for capilairy (rdc), for hydrostatic equilibrium (rdh)
  rdh = rdc = rd_spring = rd_fall = NULL
  rdh_spring = rdh_fall = req_spring_depth_day = req_fall_depth_day = NULL
  rsl = score = NULL
  
  # Load in the datasets
  crops.obic <- as.data.table(OBIC::crops.obic)
  setkey(crops.obic, crop_code)
  soils.obic <- as.data.table(OBIC::soils.obic)
  setkey(soils.obic, soiltype)
  season.obic <- as.data.table(OBIC::season.obic)
  setkey(season.obic, landuse)

  # Check inputs
  arg.length <- max(length(A_CLAY_MI), length(A_SILT_MI), length(B_LU_BRP), length(B_SOILTYPE_AGR), 
                    length(B_GWL_GLG), length(B_GWL_GHG), length(B_GWL_ZCRIT))
  checkmate::assert_numeric(A_CLAY_MI, lower = 0, upper = 100, any.missing = FALSE, len = arg.length)
  checkmate::assert_numeric(A_SILT_MI, lower = 0, upper = 100, any.missing = FALSE, len = arg.length)
  checkmate::assert_numeric(B_LU_BRP, any.missing = FALSE, min.len = 1, len = arg.length)
  checkmate::assert_subset(B_LU_BRP, choices = unique(crops.obic$crop_code), empty.ok = FALSE)
  checkmate::assert_character(B_SOILTYPE_AGR, any.missing = FALSE, min.len = 1, len = arg.length)
  checkmate::assert_subset(B_SOILTYPE_AGR, choices = unique(soils.obic$soiltype), empty.ok = FALSE)
  checkmate::assert_numeric(B_GWL_GLG, lower = 0, any.missing = FALSE, len = arg.length)
  checkmate::assert_numeric(B_GWL_GHG, lower = 0, any.missing = FALSE, len = arg.length)
  checkmate::assert_true(all(B_GWL_GHG < B_GWL_GLG))
  checkmate::assert_numeric(B_GWL_ZCRIT, lower = 0, upper = 400, any.missing = FALSE, len = arg.length)

  # Collect in data table
  dt <- data.table(id = 1:arg.length,
                   A_CLAY_MI = A_CLAY_MI,
                   A_SILT_MI = A_SILT_MI,
                   B_LU_BRP = B_LU_BRP,
                   B_SOILTYPE_AGR = B_SOILTYPE_AGR,
                   B_GWL_GLG = B_GWL_GLG,
                   B_GWL_GHG = B_GWL_GHG,
                   B_GWL_ZCRIT = B_GWL_ZCRIT)
  
  # merge with OBIC crop and soil table
  dt <- merge(dt, crops.obic[, list(crop_code, crop_waterstress, crop_season)], 
              by.x = "B_LU_BRP", by.y = "crop_code")
  dt <- merge(dt, soils.obic[, list(soiltype, soiltype.m)], by.x = "B_SOILTYPE_AGR", by.y = "soiltype")
  dt <- merge(dt, season.obic, by.x = c('crop_season','soiltype.m'), by.y = c('landuse', 'soiltype.m'))
  
  ## determine workability key numbers

    # new parameters to be added
    cols <- c('gws_sub_workingdepth','spring_depth')
    
    # sandy soils with variable silt content
    dt[soiltype.m == 'zand' & A_SILT_MI < 10, c(cols) := list(45,35)]
    dt[soiltype.m == 'zand' & A_SILT_MI >= 10 & A_SILT_MI < 20, c(cols) := list(55,30)]
    dt[soiltype.m == 'zand' & A_SILT_MI >= 20, c(cols) := list(60,30)]
      
    # loess and peat soils
    dt[soiltype.m == 'loess',c(cols) := list(65,12)]
    dt[soiltype.m == 'veen',c(cols) := list(55,22)]
    
    # clay soils
    dt[soiltype.m == 'klei' & A_CLAY_MI < 12, c(cols) := list(85,12)]
    dt[soiltype.m == 'klei' & A_CLAY_MI >= 12 & A_CLAY_MI < 17, c(cols) := list(85,12)]
    dt[soiltype.m == 'klei' & A_CLAY_MI >= 17 & A_CLAY_MI < 25, c(cols) := list(75,15)]
    dt[soiltype.m == 'klei' & A_CLAY_MI >= 25 & A_CLAY_MI < 35, c(cols) := list(65,15)]
    dt[soiltype.m == 'klei' & A_CLAY_MI >= 35, c(cols) := list(45,15)]


  # Overwrite spring working depth for perennial crops
  crops.p <- c('boomteelt', 'overig boomteelt', 'groot fruit','grasland zonder herinzaai', 'grasland met herinzaai')
  dt[crop_waterstress %in% crops.p,spring_depth := 0]
  
  # test 1: desired groundwater depth for work under hydrostatic equilibrium (abbreviated as rdh) in spring and fall
  dt[, rdh_spring := gws_sub_workingdepth+spring_depth]
  dt[, rdh_fall := gws_sub_workingdepth]
  
  # test 2: At what groundwater level is  capillary rise lower than evaporation (<2mm/d), required depth capilairy abbreviated as rdc
  dt[, rdc := B_GWL_ZCRIT]

  # Choose lowest required depth as required depth
  dt[,rd_spring := fifelse(rdh_spring <= rdc, rdh_spring,rdc)]
  dt[,rd_fall   := fifelse(rdh_fall   <= rdc, rdh_fall,  rdc)]
  
# Calculate season length -----
  
  # Calculate the day on which the desired water depth is reached for spring and fall
  
  # Spring and fall
  dt[rd_spring >= B_GWL_GHG & rd_spring <= B_GWL_GLG,
     req_spring_depth_day := round(138-(asin((-rd_spring-0.5*(-B_GWL_GHG-B_GWL_GLG))/(0.5*(-B_GWL_GHG+B_GWL_GLG)))/0.0172024))]
  dt[rd_fall >= B_GWL_GHG & rd_fall <= B_GWL_GLG,
     req_fall_depth_day := round(138-(asin((-rd_fall-0.5*(-B_GWL_GHG-B_GWL_GLG))/(0.5*(-B_GWL_GHG+B_GWL_GLG)))/0.0172024))]
  
  # if highest groundwater level is deeper than required depth, soil is always workable
  # if lowest groundwater level is higher than required depth, soil is never workable
  # then required_depth_day is set to 228 (which is 15 aug/GLG) so season length will be 0
  dt[rd_spring < B_GWL_GHG, req_spring_depth_day := 1] 
  dt[rd_spring >= B_GWL_GLG, req_spring_depth_day := 228] 
  dt[rd_fall < B_GWL_GHG, req_fall_depth_day := 1] 
  dt[rd_fall > B_GWL_GLG, req_fall_depth_day := 228]
  
  # Calculate the number of days deficit compared to ideal situation, always above zero
  dt[,early_season_day_deficit := pmax(0,req_days_pre_glg-(228-req_spring_depth_day))]
  dt[,late_season_day_deficit := pmax(0,req_days_post_glg-(228-req_fall_depth_day))]
  
  # Calculate relative season length
  dt[,rsl := (total_days-late_season_day_deficit-early_season_day_deficit)/total_days]
  
  # Calculate percentage yield loss non-grassland by sub-optimal season length
  if(calcyieldloss == TRUE){
    
    # add yield loss per category
    dt[derving == 'zaai groenten' , yl := 538 * rsl^2-1144 * rsl + 606]
    dt[derving == 'zomergranen'   , yl := 232 * rsl^2- 475 * rsl + 243]
    dt[derving == 'plant groenten', yl := 392 * rsl^2- 785 * rsl + 393]
    dt[derving == 'wintergranen'  , yl :=(232 * rsl^2- 475 * rsl + 243)*0.85/2]
    dt[derving == 'boomteelt'     , yl :=(538 * rsl^2-1144 * rsl + 606)/2]
    dt[derving == 'overig'        , yl := 100 * rsl^2- 200 * rsl + 100]
    
    # # helper functions to determine yield loss in grass given soil type
    ylveen <- approxfun(x = c(1, 0.9, 0.8, 0.6, 0.55, 0.43, 0.22, 0.08, 0),
                        y = c(23, 25, 28, 36, 38, 41, 54, 59, 100), method = 'linear',
                        yleft = NA_integer_, yright = NA_integer_)
    ylsoil <- approxfun(x = c(1, 0.9, 0.8, 0.6, 0.55, 0.43, 0.22, 0.08, 0),
                        y = c(23, 25, 29, 41, 43, 51, 68, 72, 100), method = 'linear',
                        yleft = NA_integer_, yright = NA_integer_)
 
    # add yield reduction for grassland
    dt[derving == 'grasland' & soiltype.m == 'veen', yl := ylveen(rsl)]
    dt[derving == 'grasland' & !soiltype.m == 'veen', yl := ylsoil(rsl)]

    # Calculate yield fraction, always above zero
    dt[,yield := pmax(0, 1 - 0.01 * yl)] 
    dt[derving == 'grasland', yield := evaluate_logistic(x = yield, b = 16, x0 = 0.5, 
                                                         v = 0.5,increasing = TRUE)]
    
  }
    
  
  # setorder
  setorder(dt, id)
  
  # Return output
  if(calcyieldloss == TRUE){
    
    # return yield decline
    value <- dt[,yield]
    
  } else {
    
    # return relative length season
    value <- dt[,rsl]  
  }
  
  
  # return value
  return(value)
}

#' Calculate indicator for workability
#'
#' This function calculates the indicator for the workability of the soil expressed as the period in which the soil can be worked without
#' inflicting structural damage that cannot be restored by the regular management on the farm.
#'  
#' @param D_WO (numeric) The value of the relative (workable) season length calculated by \code{\link{calc_workability}}
#' @param B_LU_BRP (numeric) The crop code from the BRP
#'  
#' @export
ind_workability <- function(D_WO, B_LU_BRP) {
  
  # add visual bindings
  id = arg.length = crop_code = crop_season = rsl = . = NULL
  
  # Load in the datasets
  crops.obic <- as.data.table(OBIC::crops.obic)
  setkey(crops.obic, crop_code)
  
  # length of inputs
  arg.length <- max(length(D_WO), length(B_LU_BRP))
  
  # Check inputs
  checkmate::assert_numeric(D_WO, lower = 0, upper = 1, any.missing = FALSE)
  checkmate::assert_numeric(B_LU_BRP, any.missing = FALSE, min.len = 1, len = arg.length)
  checkmate::assert_subset(B_LU_BRP, choices = unique(crops.obic$crop_code), empty.ok = FALSE)
  
  # Form data table
  dt <- data.table(id = 1:arg.length,
                   rsl = D_WO,
                   B_LU_BRP = B_LU_BRP)
  
  # Merge crop_season into data table
  dt <- merge.data.table(dt, crops.obic[,.(crop_code, crop_season)], by.x = 'B_LU_BRP', by.y = 'crop_code')
  
  # evaluate relative season length
  dt[!grepl('^beweid', crop_season),score := evaluate_logistic(x = rsl, b = 15, x0 = 0.75, v = 1, increasing = TRUE)]
  dt[grepl('^beweid', crop_season),score := evaluate_logistic(x = rsl, b = 9, x0 = 0.5, v = 1, increasing = TRUE)]
  
  # overwrite score when rsl = 1
  dt[rsl == 1, score := 1]
  
  # setorder
  setorder(dt, id)
  
  # Return score
  score <- dt[,score]
  
  return(score)
}
